lib/Dialect/FIRRTL/Transforms/EmitOMIR.cpp

```diff
@@ -1208,6 +1242,44 @@ hw::InnerRefAttr EmitOMIRPass::getInnerRefTo(FModuleLike module,
                          });
 }
 
+FIRRTLType EmitOMIRPass::getTypeOf(Operation *op) {
+  assert(op->getNumResults() == 1 &&
+         op->getResult(0).getType().isa<FIRRTLType>() &&
+         "op must have a single FIRRTLType result");
+  return op->getResult(0).getType().cast<FIRRTLType>();
+}
+
+FIRRTLType EmitOMIRPass::getTypeOf(FModuleLike mod, size_t portIdx) {
+  Type portType = mod.getPortType(portIdx);
+  assert(portType.isa<FIRRTLType>() && "port must have a FIRRTLType");
+  return portType.cast<FIRRTLType>();
+}
+
+// Constructs a reference to a field of an aggregate FIRRTLType with a fieldID,
+// and appends it to result. If fieldID is 0, meaning it does not reference a
+// field of an aggregate FIRRTLType, this is a no-op.
+void EmitOMIRPass::addFieldID(FIRRTLType type, unsigned fieldID,
+                              SmallVectorImpl<char> &result) {
+  while (fieldID)
+    TypeSwitch<FIRRTLType>(type)
+        .Case<FVectorType>([&](FVectorType vector) {
+          size_t index = vector.getIndexForFieldID(fieldID);
+          type = vector.getElementType();
+          fieldID -= vector.getFieldID(index);
+          result.push_back('[');
+          Twine(index).toVector(result);
+          result.push_back(']');
+        })
+        .Case<BundleType>([&](BundleType bundle) {
+          size_t index = bundle.getIndexForFieldID(fieldID);
+          StringRef name = bundle.getElement(index).name;
+          type = bundle.getElementType(index);
+          fieldID -= bundle.getFieldID(index);
+          result.push_back('.');
+          result.append(name.begin(), name.end());
+        });
```

Comment:
Not sure if we care about this, but could this hang if the input OMIR is badly formed and has fieldID on a non-aggregate type?

Reply:
hmmm yes this while loop is a little scary. I could add an assertion that if fieldID is nonzero, the type is aggregate? or is there something cleaner to make it safe?